[Frontend] Set server's maximum number of generated tokens using generation_config.json

mhendrey opened the pull request:
This PR adds in the ability to specify the `max_new_tokens` entry in the generation_config.json file. Setting this parameter will act as the server's maximum number of generated tokens any given request can have. 

If a user does not specify a `max_tokens` in their request, then the minimum of `max_new_tokens` and (`max_model_len` - prompt_tokens) will be used.  Current behavior just uses the `max_model_len` - prompt_tokens as a default. For large context window models, e.g., Llama-3.1, this is 128K.

If a user does specify a `max_tokens`, then the minimum of `max_tokens`, `max_new_tokens`, or `max_model_len` - prompt_tokens will be used. Current behavior allows user to specify `max_tokens` larger than is physically allowed by the context window, though it will throw an error.  Now it would quietly override user's requested `max_tokens`.

Still need to add some testing, but would appreciate some pointers to where the tests should be added. @DarkLight1337 suggested it may not be necessary to update the `vllm/entrypoints/llm.py`, but I've left that in for now until final determination can be made.

FIX #[11976](https://github.com/vllm-project/vllm/issues/11976)

ShangmingCai committed:
[Bugfix] Fix num_heads value for simple connector when tp enabled (#12074)

Signed-off-by: Shangming Cai <caishangming@linux.alibaba.com>
Signed-off-by: Matthew Hendrey <matthew.hendrey@gmail.com>

youkaichao committed:
[torch.compile] fix sym_tensor_indices (#12191)

Signed-off-by: youkaichao <youkaichao@gmail.com>
Signed-off-by: Matthew Hendrey <matthew.hendrey@gmail.com>

mhendrey committed:
Adding max_new_tokens support to generation_config.json

ModelConfig.get_diff_sampling_params() now allows for reading the "max_new_tokens"
if its specified in the generation_config.json file. This follows Huggingface's
naming convention for the variable that specifies the maximum number of generated tokens.
This gets renamed to "max_tokens" to follow the naming convention used by vLLM for the same functionality.

Signed-off-by: Matthew Hendrey <matthew.hendrey@gmail.com>

mhendrey committed:
Changed default_max_tokens to server_max_tokens

Previously the default_max_tokens was the (max_model_len - prompt_tokens),
but now the server_max_tokens = min(
  max_model_len - prompt_tokens,
  max_tokens if set in generation_config.json
)

Signed-off-by: Matthew Hendrey <matthew.hendrey@gmail.com>

mhendrey committed:
Renamed default_max_tokens to server_max_tokens

server_max_tokens is the minimum between architectural limitations,
which was the default_max_tokens, and max_new_tokens set in generation_config.json

Signed-off-by: Matthew Hendrey <matthew.hendrey@gmail.com>

mhendrey committed:
Removed the float("inf") bug

I have int's so that wasn't good. I could have gone with 2**64 or something, but
didn't like the idea of some hardcoded value. So changed a logic just a touch.

Signed-off-by: Matthew Hendrey <matthew.hendrey@gmail.com>

mhendrey committed:
Renamed default_max_tokens to server_max_tokens

Also added in setting server_max_tokens to the minimum of context window - prompt and the value of max_new_tokens set in generation_config.json

Signed-off-by: Matthew Hendrey <matthew.hendrey@gmail.com>

mhendrey committed:
Rearranged lines to make the changes with existing as small as possible

Signed-off-by: Matthew Hendrey <matthew.hendrey@gmail.com>

mhendrey committed:
Limit generated tokens by server's max_tokens setting when available

server_max_tokens is set either by architectural limits (context_window - prompt_tokens) or max_new_tokens value set in generation_config.json

Signed-off-by: Matthew Hendrey <matthew.hendrey@gmail.com>

mhendrey committed:
Changed syntax to pass format.sh tests

Signed-off-by: Matthew Hendrey <matthew.hendrey@gmail.com>

hmellor committed:
Move linting to `pre-commit` (#11975)

Signed-off-by: Harry Mellor <19981378+hmellor@users.noreply.github.com>
Signed-off-by: Matthew Hendrey <matthew.hendrey@gmail.com>

terrytangyuan committed:
[DOC] Fix typo in docstring and assert message (#12194)

Signed-off-by: Yuan Tang <terrytangyuan@gmail.com>
Signed-off-by: Matthew Hendrey <matthew.hendrey@gmail.com>

terrytangyuan committed:
[DOC] Add missing docstring in LLMEngine.add_request() (#12195)

Signed-off-by: Yuan Tang <terrytangyuan@gmail.com>
Signed-off-by: Matthew Hendrey <matthew.hendrey@gmail.com>

terrytangyuan committed:
[Bugfix] Fix incorrect types in LayerwiseProfileResults (#12196)

Signed-off-by: Yuan Tang <terrytangyuan@gmail.com>
Signed-off-by: Matthew Hendrey <matthew.hendrey@gmail.com>

Isotr0py committed:
[Model] Add Qwen2 PRM model support (#12202)

Signed-off-by: Isotr0py <2037008807@qq.com>
Signed-off-by: Matthew Hendrey <matthew.hendrey@gmail.com>

DarkLight1337 committed:
[Core] Interface for accessing model from `VllmRunner` (#10353)

Signed-off-by: DarkLight1337 <tlleungac@connect.ust.hk>
Signed-off-by: Matthew Hendrey <matthew.hendrey@gmail.com>

youkaichao committed:
[misc] add placeholder format.sh (#12206)

Signed-off-by: youkaichao <youkaichao@gmail.com>
Signed-off-by: Matthew Hendrey <matthew.hendrey@gmail.com>

DarkLight1337 committed:
[CI/Build] Remove dummy CI steps (#12208)

Signed-off-by: DarkLight1337 <tlleungac@connect.ust.hk>
Signed-off-by: Matthew Hendrey <matthew.hendrey@gmail.com>

DarkLight1337 committed:
[CI/Build] Make pre-commit faster (#12212)

Signed-off-by: DarkLight1337 <tlleungac@connect.ust.hk>
Signed-off-by: Matthew Hendrey <matthew.hendrey@gmail.com>

DarkLight1337 committed:
[Model] Upgrade Aria to transformers 4.48 (#12203)

Signed-off-by: DarkLight1337 <tlleungac@connect.ust.hk>
Signed-off-by: Matthew Hendrey <matthew.hendrey@gmail.com>

youkaichao committed:
[misc] print a message to suggest how to bypass commit hooks (#12217)

Signed-off-by: youkaichao <youkaichao@gmail.com>
Signed-off-by: Matthew Hendrey <matthew.hendrey@gmail.com>

youkaichao committed:
[core][bugfix] configure env var during import vllm (#12209)

Signed-off-by: youkaichao <youkaichao@gmail.com>
Signed-off-by: Matthew Hendrey <matthew.hendrey@gmail.com>

heheda12345 committed:
[V1] Remove `_get_cache_block_size` (#12214)

Signed-off-by: Chen Zhang <zhangch99@outlook.com>
Signed-off-by: Matthew Hendrey <matthew.hendrey@gmail.com>

wangxiyuan committed:
[Misc] Pass `attention` to impl backend (#12218)

Signed-off-by: wangxiyuan <wangxiyuan1007@gmail.com>
Signed-off-by: Matthew Hendrey <matthew.hendrey@gmail.com>

DarkLight1337 committed:
[Bugfix] Fix `HfExampleModels.find_hf_info` (#12223)

Signed-off-by: DarkLight1337 <tlleungac@connect.ust.hk>
Signed-off-by: Matthew Hendrey <matthew.hendrey@gmail.com>

heheda12345 committed:
[CI] Pass local python version explicitly to pre-commit mypy.sh (#12224)

Signed-off-by: Chen Zhang <zhangch99@outlook.com>
Signed-off-by: Matthew Hendrey <matthew.hendrey@gmail.com>

DarkLight1337 commented:
You can add tests under `tests/entrypoints/openai`

mhendrey committed:
Added tests to check max_tokens is properly set

Signed-off-by: Matthew Hendrey <matthew.hendrey@gmail.com>

mhendrey commented:
I added to the existing tests/entrypoints/openai/test_serving_chat.py's test_serving_chat_should_set_correct_max_tokens().

Still not quite sure if the vllm/entrypoints/llm.py needs to be edited. Is this only used for people that are running locally? If there is a generation_config.json file, do we want that to limit the number of generated tokens if `max_new_tokens` is specified?  Or maybe this should act more like a default value for a single user?

DarkLight1337 commented:
`LLM` class is only used in offline inference. There is no need to update it because `get_diff_sampling_param` already sets the correct value from `generation_config.json`.

DarkLight1337 commented:
By the way, can you fix the Git history? Might need to undo some merge commits.

mhendrey commented:
Thanks for the insight on LLM. I'll undo the changes on that file and work on resolving the merge conflicts. Probably get to that tomorrow.

mhendrey committed:
Merge branch 'server_max_tokens'

Signed-off-by: Matthew Hendrey <matthew.hendrey@gmail.com>

mhendrey committed:
Mucked up the rebasing.  Fixing that now.

These files should not be different from what's in main

Signed-off-by: Matthew Hendrey <matthew.hendrey@gmail.com>

mhendrey commented:
Fumbling around a little bit with rebasing. I think I got it and reopening this PR.

DarkLight1337 commented on vllm/entrypoints/openai/serving_completion.py:
```diff
@@ -115,17 +115,24 @@ async def create_completion(
                 default_sampling_params = (
                     self.model_config.get_diff_sampling_param())
+
+                # Limit set by architecture or value in generation_config.json
+                if "max_tokens" in default_sampling_params:
+                    server_max_tokens = min(
+                        server_max_tokens,
+                        default_sampling_params["max_tokens"])
```
Since `default_sampling_params` is also passed to `request.to_beam_search_params` and `request.to_sampling_params`, let's handle this inside those methods instead.

mhendrey committed:
Reverting the serving_chat & serving_completion back and putting all the logic in protocol.py

Signed-off-by: Matthew Hendrey <matthew.hendrey@gmail.com>

mhendrey committed:
Didn't quite revert back. Deleting empty line from both

Signed-off-by: Matthew Hendrey <matthew.hendrey@gmail.com>

DarkLight1337 commented on vllm/entrypoints/openai/protocol.py:
```diff
@@ -380,13 +380,19 @@ def to_beam_search_params(
         n = self.n if self.n is not None else 1
 
+        # Use minimum of context window, user request & server limit.
+        max_tokens_choices = [
+            val for val in (default_max_tokens, max_tokens,
+                            default_sampling_params.get("max_tokens", None))
+            if val is not None
+        ]
+        max_tokens = min(max_tokens_choices)
```
```suggestion
        max_tokens = min(
            val for val in (default_max_tokens, max_tokens,
                            default_sampling_params.get("max_tokens", None))
            if val is not None
        )
```
Can compute this in one statement

mhendrey replied:
I've made the suggested change and updated the engine args description for generation-config. Kicking myself for not making the one-liner. Sorry about that.

mhendrey committed:
Changed to using one-liner and edited engine arg for generation-config

Uses a one-liner for taking the min of the user's max_tokens, the context window - prompt tokens, and value set in generation_config.json.

Updated the generation-config argument "help" to describe what happens when max_new_tokens is specified.

mhendrey committed:
Merge branch 'vllm-project:main' into main

mhendrey committed:
Converted to a one-liner for taking minimum value & added to generation-config argument description

Simplified to using just a one-liner to take the minimum of the user's max_tokens, context window - prompt tokens, or value set in generation_config.json file.

Updated the command-line argument description for generation-config to reflect that setting max_new_tokens sets a server-wide limit for number of output tokens.

Merge branch 'server_max_tokens'

DarkLight1337 reviewed (approved):
Looks good, thanks for implementing this!

mhendrey commented:
Really appreciate the comments/feedback you have given to make this possible. Thank you.

It looks like it is failing some AMD tests (FAILED engine/test_multiproc_workers.py::test_local_workers_clean_shutdown - ...). I hadn't touched that part of the code.  Is this something that I need to try and resolve?

DarkLight1337 commented:
I'm getting someone to force-merge this.

DarkLight1337 commented:
> FYI, this pr was not mentioned in the release notes. However, unless I'm misunderstanding something, the combination of
> https://github.com/vllm-project/vllm/pull/12622 and this PR (https://github.com/vllm-project/vllm/pull/12242) may truncate a user's expectation of the output tokens since it defaults to the minimum.
> 
> e.g. [Qwen/Qwen2.5-32B](https://huggingface.co/Qwen/Qwen2.5-32B/blob/main/generation_config.json)
> 
> Please correct me if I'm wrong!

Yes, that is correct. Do you prefer raising an error in those cases?

JGSweets commented:
> Yes, that is correct. Do you prefer raising an error in those cases?

I'm curious of how OpenAI would handle it.

I personally could see it being a 400 Bad Request since the user is requesting output that is not allowed by the service. Or should raising the 400 be optionally set by the service?

Open to thoughts.

DarkLight1337 commented:
I think raising 400 is fine in this case as well. Are you free to open a PR?
